src/include/OpenImageIO/imageio.h

```diff
@@ -2787,10 +2791,45 @@ class OIIO_API ImageOutput {
     std::unique_ptr<Impl, decltype(&impl_deleter)> m_impl;
 
     void append_error(string_view message) const; // add to m_errmessage
+
+    /// declare a friend heapsize definition
+    template <typename T> friend inline size_t pvt::heapsize(const T&);
 };
 
 
 
+// Memory tracking functions
+
+/// Specializes the base memory tracking functions from memory.h.
+namespace pvt {
+// heapsize specialization for `ImageSpec`
+template <> inline size_t heapsize<ImageSpec>(const ImageSpec& is)
+{
+    size_t size = heapsize(is.channelformats);
+    size += heapsize(is.channelnames);
+    size += heapsize(is.extra_attribs);
+    return size;
+}
```

Comment:
What I think I would recommend for these type-specialized cases is that in the public header file, we should only declare the specialization:

```
// pvt::heapsize specialization for `ImageSpec`
template <> OIIO_API size_t pvt::heapsize<ImageSpec>(const ImageSpec& is);
```

Then in just one module, formatspec.cpp in this case, define the implementation:

```
template <> size_t heapsize<ImageSpec>(const ImageSpec& is);
{
    size_t size = heapsize(is.channelformats);
    size += heapsize(is.channelnames);
    size += heapsize(is.extra_attribs);
    return size;
}
```

I think this will (a) keep the clutter out of the header files; (b) make it so that the implementation is only compiled ONCE, not in every module that sees the header file; (c) hide the implementation details and allow us to change them without forcing a recompile of the whole codebase.

I'm only speaking of the OIIO-type-specialized ones, like `heapsize<ImageSpec>`; the broadly generic ones like `heapsize<vector<T>>` of course just stay in the memory.h header.

Reply:
Thanks @lgritz , will implement all your remarks right now !